lib/auth/join_gcp.go

```diff
@@ -61,24 +65,36 @@ func (a *Server) checkGCPJoinRequest(ctx context.Context, req *types.RegisterUsi
 
 func checkGCPAllowRules(token *types.ProvisionTokenV2, claims *gcp.IDTokenClaims) error {
 	compute := claims.Google.ComputeEngine
+	// unmatchedLocation is true if the location restriction is set and the "google.compute_engine.zone"
+	// claim is not present in the IDToken. This happens when the joining node is not a GCE VM.
+	unmatchedLocation := false
 	// If a single rule passes, accept the IDToken.
 	for _, rule := range token.Spec.GCP.Allow {
 		if !slices.Contains(rule.ProjectIDs, compute.ProjectID) {
 			continue
 		}
+
+		if len(rule.ServiceAccounts) > 0 && !slices.Contains(rule.ServiceAccounts, claims.Email) {
+			continue
+		}
+
 		if len(rule.Locations) > 0 && !slices.ContainsFunc(rule.Locations, func(location string) bool {
 			return isGCPZoneInLocation(location, compute.Zone)
 		}) {
-			continue
-		}
-		if len(rule.ServiceAccounts) > 0 && !slices.Contains(rule.ServiceAccounts, claims.Email) {
+			unmatchedLocation = true
 			continue
 		}
 
 		// All provided rules met.
 		return nil
 	}
 
+	// If the location restriction is set and the "google.compute_engine.zone" claim is not present in the IDToken,
+	// return a more specific error message.
+	if unmatchedLocation && compute.Zone == "" {
+		return trace.CompareFailed("id token %q claim is empty and didn't match the %q. "+
```

Comment:
Why here the format string is always used to insert the same value?

Reply:
I used the format string to avoid having to escape `"`

Reply:
You could also use a raw string (with `` ` ``) to do the same thing.